### PR TITLE
FEAT#2917: CudfOnRay Indexing

### LINKED ISSUE #2917: Initial Merge of MODIN-GPU
Refer to issue #2538 for context. An initial version of MODIN-GPU can be merged, but indexing support must be added first. This inlcudes iloc, loc, support, and requires the code to pass the tests inside test_indexing.py.

### PR DESCRIPTION
**Pull request by kvu35**
## What do these changes do?
This is the first PR from the MODIN-GPU project to the parent MODIN. This PR adds the following functionality that has been developed so far by the gatech-sysml MODIN-GPU team:
- cudfQueryCompiler, cudfOnRayFrame, cudfOnRayPartition, cudfOnRayPartitionManager.
- Basic IO, read_csv
- from_pandas functionalities, basic dataframe construction.
- Indexing (still a todo).

- [x] commit message follows format outlined
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #2917 
- [ ] tests added and passing

### CODE REVIEW

**Comment by pastranakike** [comment]
@pastranakike

**Comment by pastranakike** [comment]
Last commit fixed indexing and passed over 10 tests. I cleaned the QC commenting out all the methods that we are not supporting yet. Also, I included commenting for all the modin-gpu relevant layers except the partition (as it seems pretty straightforward). Lastly, this commits passes black/flake8.

**Comment by devin-petersohn** [comment]
@vnlitvinov Can you help me understand why the cloud tests is failing?

**Comment by vnlitvinov** [comment]
Something really weird is going on...

Relevant error:
```
2021-03-26T21:57:37.4590987Z E           ========= Remote Traceback (1) =========
2021-03-26T21:57:37.4591373Z E           Traceback (most recent call last):
2021-03-26T21:57:37.4592251Z E             File "/usr/share/miniconda/envs/modin/lib/python3.7/site-packages/rpyc/core/protocol.py", line 320, in _dispatch_request
2021-03-26T21:57:37.4592915Z E               res = self._HANDLERS[handler](self, *args)
2021-03-26T21:57:37.4593775Z E             File "/usr/share/miniconda/envs/modin/lib/python3.7/site-packages/rpyc/core/protocol.py", line 593, in _handle_call
2021-03-26T21:57:37.4594396Z E               return obj(*args, **dict(kwargs))
2021-03-26T21:57:37.4595220Z E             File "/usr/share/miniconda/envs/modin/lib/python3.7/site-packages/rpyc/core/service.py", line 160, in getmodule
2021-03-26T21:57:37.4595841Z E               return __import__(name, None, None, "*")
2021-03-26T21:57:37.4597757Z E             File "/home/runner/work/modin/modin/modin/experimental/cloud/rpyc_proxy.py", line 24, in <module>
2021-03-26T21:57:37.4598408Z E               from .meta_magic import _LOCAL_ATTRS, RemoteMeta, _KNOWN_DUALS
2021-03-26T21:57:37.4599063Z E             File "/home/runner/work/modin/modin/modin/experimental/cloud/meta_magic.py", line 19, in <module>
2021-03-26T21:57:37.4601005Z E               from modin.data_management.factories import REMOTE_ENGINES
2021-03-26T21:57:37.4602015Z E             File "/home/runner/work/modin/modin/modin/data_management/factories/__init__.py", line 14, in <module>
2021-03-26T21:57:37.4602742Z E               from . import factories
2021-03-26T21:57:37.4603554Z E             File "/home/runner/work/modin/modin/modin/data_management/factories/factories.py", line 19, in <module>
2021-03-26T21:57:37.4604988Z E               from modin.engines.base.io import BaseIO
2021-03-26T21:57:37.4605612Z E             File "/home/runner/work/modin/modin/modin/engines/base/io/__init__.py", line 15, in <module>
2021-03-26T21:57:37.4606416Z E               from modin.engines.base.io.text.csv_dispatcher import CSVDispatcher
2021-03-26T21:57:37.4607266Z E             File "/home/runner/work/modin/modin/modin/engines/base/io/text/csv_dispatcher.py", line 14, in <module>
2021-03-26T21:57:37.4608103Z E               from modin.engines.base.io.text.text_file_dispatcher import (
2021-03-26T21:57:37.4608937Z E             File "/home/runner/work/modin/modin/modin/engines/base/io/text/text_file_dispatcher.py", line 23, in <module>
2021-03-26T21:57:37.4609599Z E               from modin.pandas import GPU_MANAGERS
2021-03-26T21:57:37.4610291Z E             File "/home/runner/work/modin/modin/modin/pandas/__init__.py", line 171, in <module>
2021-03-26T21:57:37.4610868Z E               from .dataframe import DataFrame
2021-03-26T21:57:37.4611434Z E             File "/home/runner/work/modin/modin/modin/pandas/dataframe.py", line 54, in <module>
2021-03-26T21:57:37.4611989Z E               from .series import Series
2021-03-26T21:57:37.4612513Z E             File "/home/runner/work/modin/modin/modin/pandas/series.py", line 41, in <module>
2021-03-26T21:57:37.4613162Z E               from .base import BasePandasDataset, _ATTRS_NO_LOOKUP
2021-03-26T21:57:37.4613785Z E             File "/home/runner/work/modin/modin/modin/pandas/base.py", line 2839, in <module>
2021-03-26T21:57:37.4614537Z E               from modin.experimental.cloud.meta_magic import make_wrapped_class
2021-03-26T21:57:37.4616013Z E           ImportError: cannot import name 'make_wrapped_class' from 'modin.experimental.cloud.meta_magic' (/home/runner/work/modin/modin/modin/experimental/cloud/meta_magic.py)
```

As you can see, it is not even starting to test cloud features...

**Comment by devin-petersohn** [comment]
I did not know where to begin on debugging this 😄

**Comment by vnlitvinov** [comment]
Let's start with a rebase over current `master`, maybe it was a one-off issue with CI?..

**Comment by vnlitvinov** [comment]
> I did not know where to begin on debugging this 😄

yet more reasons for https://github.com/modin-project/modin/issues/2934 to finally happen btw 😆

**Comment by pastranakike** [comment]
I see that the test is failing with:
```
File "/home/runner/work/modin/modin/modin/engines/base/io/text/text_file_dispatcher.py", line 23, in <module>
2021-03-26T21:57:37.4609599Z E               from modin.pandas import GPU_MANAGERS
```
should we move the definition of the GPU_MANAGERS array? I do not know how the gpu engine is interfering with that test. Is there something that me or @kvu35 can do to help?

**Comment by devin-petersohn** [comment]
@pastranakike That would make sense, let's create a separate file in modin/engines/ray/cudf_on_ray/io to override `text_file_dispatcher` or monkey patch in the changes to `build_partitions` and `_launch_tasks`. 

I think we should probably remove the `GPU_MANAGERS` from the modin/pandas module anyway because it is specific to the cudf on Ray implementation.

**Comment by pastranakike** [comment]
sgtm! Let me rebase with master and monkey patch those methods.

File: modin/engines/ray/cudf_on_ray/io.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+)
+from modin.engines.ray.cudf_on_ray.frame.partition import (
+    cuDFOnRayFramePartition,
+)
+
+
+from modin.engines.ray.task_wrapper import RayTask
+from modin.backends.cudf.parser import cuDFCSVParser
+
+
```
Comment (devin-petersohn) [inline modin/engines/ray/cudf_on_ray/io.py:29]:
```suggestion
class GPUCSVDispatcher(CSVDispatcher):
    def build_parititon(cls, ...):
        # Implementation here
        pass
```

**Comment by devin-petersohn** [comment]
Thanks @YarShev for restarting TeamCity CI.

**Comment by YarShev** [comment]
@devin-petersohn , no problem!

**Review by devin-petersohn — APPROVED** [review]
Thanks @kvu35, LGTM!

File: modin/engines/ray/cudf_on_ray/frame/data.py [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+                # Column widths for slice are calculated as the length of the slice
+                # on the partition. Often this will be the same length as the current
+                # length, but sometimes it is different, thus the extra calculation.
+                new_col_widths = [
+                    len(range(*idx.indices(self._column_widths[p])))
+                    for p, idx in col_partitions_list.items()
+                ]
+                # Use the slice to calculate the new columns
+                new_columns = self.columns[col_numeric_idx]
+                assert sum(new_col_widths) == len(
```
Comment (jbrockmendel) [inline modin/engines/ray/cudf_on_ray/frame/data.py:175]: the code around here is really similar but not identical to what we have in PandasDataFrame.mask (now take_2d_labels_or_positional).  are the differences intentional?  or could we share some of it?

**Pull request merged**